### PR TITLE
feat: Place limit first pass implementation

### PR DESCRIPTION
**Pull request by crnbarr93**
## What is the purpose of the change:
These changes introduce a first pass implementation of placing limit orders. Designs are finalized for the placing of buy and sell orders on the orderbook, however this PR does not include all design changes. 

The primary reason for this PR is to allow future updates to branch off the feature branch with the changes made on the `connor/place-limit` branch.

**This is not a finished implementation**

### Linear Task

[LIM-89: Place Limit Order](https://linear.app/osmosis/issue/LIM-89/place-limit-order)

## Brief Changelog

- Added new `TradeTool` component
- Added tabs for `TradeTool` to allow trade type selection
- Added new `PlaceLimit` component to allow for placing both a buy and sell limit order
- Added new `TokenSelectLimit` component to allow for restricted selection of denom pairs for limit orders
- Added new `usePlaceLimit` hook to manage limit order placing state
- Added new `useOrderbook` hook to fetch orderbook related info from pool store [WIP]
- Added new `LimitInput` component that swaps between fiat ($) and token inputs

### CODE REVIEW

File: packages/web/components/control/token-select-limit.tsx [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+          )}
+          {tokenSelectionAvailable && (
+            <button
+              onClick={(e) => {
+                e.stopPropagation();
+                if (tokenSelectionAvailable) {
+                  setIsSelectOpen(!isSelectOpen);
+                }
+              }}
+              className="h-[32px] rounded-2xl bg-osmoverse-800 py-1 px-3 text-body2 text-wosmongton-200"
```
Comment (fabryscript) [inline packages/web/components/control/token-select-limit.tsx:164]:
```suggestion
              className="h-8 rounded-2xl bg-osmoverse-800 py-1 px-3 text-body2 text-wosmongton-200"
```

File: packages/web/components/control/token-select-limit.tsx [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            >
+              Change
+            </button>
+          )}
+        </div>
+        <div className="align-center relative z-0 mt-[-20px] flex flex-row place-content-between items-center rounded-xl bg-osmoverse-1000 py-5 px-3 pt-10 md:justify-start">
```
Comment (fabryscript) [inline packages/web/components/control/token-select-limit.tsx:170]:
```suggestion
        <div className="align-center relative z-0 -mt-5 flex place-content-between items-center rounded-xl bg-osmoverse-1000 py-5 px-3 pt-10 md:justify-start">
```

keep in mind that the default flex direction is always row

File: packages/web/components/control/token-select-limit.tsx [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                {
+                  "opacity-40": disabled,
+                }
+              )}
+            >
+              <span className="subtitle1 text-osmoverse-300">
+                {orderDirection === OrderDirection.Bid ? "Pay with" : "Receive"}
+              </span>
+              {quoteAsset.coinImageUrl && (
+                <div className="h-[24px] w-[24px] shrink-0 rounded-full md:h-7 md:w-7">
```
Comment (fabryscript) [inline packages/web/components/control/token-select-limit.tsx:185]:
```suggestion
                <div className="h-6 w-6 shrink-0 rounded-full md:h-7 md:w-7">
```

File: packages/web/components/control/token-select-limit.tsx [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                  <Image
+                    src={quoteAsset.coinImageUrl}
+                    alt="token icon"
+                    width={24}
+                    height={24}
+                    priority
+                  />
+                </div>
+              )}
+              <div className="flex flex-row">
```
Comment (fabryscript) [inline packages/web/components/control/token-select-limit.tsx:195]:
```suggestion
              <div className="flex">
```

File: packages/web/components/place-limit-tool/index.tsx [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                    spinnerContainer: "!gap-1",
+                  }}
+                  onClick={() => swapState.inAmountInput.toggleMax()}
+                >
+                  {t("swap.MAX")}
+                </Button>
+              </Tooltip>
+            </div>
+          </div>
+          <div className="mt-3 flex place-content-between items-center">
```
Comment (fabryscript) [inline packages/web/components/place-limit-tool/index.tsx:106]: why not? `place-content-*` only applies to grid systems, not flexboxes
```suggestion
          <div className="mt-3 flex justify-between items-center">
```

File: packages/web/components/place-limit-tool/index.tsx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+              <LimitInput
+                onChange={swapState.inAmountInput.setAmount}
+                baseAsset={swapState.inAmountInput.balance!}
+                tokenAmount={swapState.inAmountInput.inputAmount}
+                price={swapState.priceState.price}
+              />
+            </div>
+          </div>
+        </div>
+        <div className="mt-3 flex place-content-between items-center text-body1">
```
Comment (fabryscript) [inline packages/web/components/place-limit-tool/index.tsx:117]: same as above

File: packages/web/components/place-limit-tool/index.tsx [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+              >{`When ${swapState.baseDenom} price is at `}</span>
+              <span
+                className={classNames(
+                  "w-full bg-transparent text-wosmongton-300 focus:outline-none "
+                )}
+              >{`$${formatPretty(swapState.priceState.price)}`}</span>
+            </div>
+          </div>
+        </div>
+        <div className="flex w-full flex-row place-content-between items-center rounded-xl border border-osmoverse-700 py-3 px-6">
```
Comment (fabryscript) [inline packages/web/components/place-limit-tool/index.tsx:133]:
```suggestion
        <div className="flex w-full justify-between items-center rounded-xl border border-osmoverse-700 py-3 px-6">
```

**Review by fabryscript — CHANGES_REQUESTED** [review]

File: packages/web/components/control/token-select-limit.tsx [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+import { OrderDirection } from "~/hooks/limit-orders";
+import { useCoinPrice } from "~/hooks/queries/assets/use-coin-price";
+import { useSwapAsset } from "~/hooks/use-swap";
+import { formatPretty } from "~/utils/formatter";
+
+export interface TokenSelectLimitProps {
+  dropdownOpen?: boolean;
+  setDropdownOpen?: (value: boolean) => void;
+  // TODO: Better typing
+  selectableAssets: ReturnType<typeof useSwapAsset>["asset"][];
```
Comment (jonator) [inline packages/web/components/control/token-select-limit.tsx:22]: If you want cleaner typing perhaps you can define a `SwapAsset` type in the use-swap file that is the return type of `useSwapAsset` (or you can define it as the return value of `useSwapAsset`)

File: packages/web/components/control/token-select-limit.tsx [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+    quoteBalance,
+    disabled,
+    orderDirection,
+  }) => {
+    const { isMobile } = useWindowSize();
+    const router = useRouter();
+    const { logEvent } = useAmplitudeAnalytics();
+
+    // parent overrideable state
+    const [isSelectOpenLocal, setIsSelectOpenLocal] = useState(false);
```
Comment (jonator) [inline packages/web/components/control/token-select-limit.tsx:61]: BTW we have a `useControllableState` hook that may assist with this

File: packages/web/components/control/token-select-limit.tsx [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+                    ) : (
+                      <div />
+                    )}
+                  </>
+                )}
+              </div>
+            </div>
+          )}
+          {tokenSelectionAvailable && (
+            <button
```
Comment (jonator) [inline packages/web/components/control/token-select-limit.tsx:157]: BTW we have a `Button` component that may have style variants that meet your needs

File: packages/web/components/drawers/token-select-drawer-limit.tsx [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+    const { t } = useTranslation();
+    const { userSettings } = useStore();
+    const { isMobile } = useWindowSize();
+    const uniqueId = useConst(() => Math.random().toString(36).substring(2, 9));
+
+    const [
+      keyboardSelectedIndex,
+      setKeyboardSelectedIndex,
+      keyboardSelectedIndexRef,
+    ] = useStateRef(0);
```
Comment (jonator) [inline packages/web/components/drawers/token-select-drawer-limit.tsx:82]: Perhaps there's a hook we can use in `react-use` package so we don't need to define our own?

File: packages/web/components/drawers/token-select-drawer-limit.tsx [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+    const { handleKeyDown: searchBarKeyDown } = useKeyActions({
+      ArrowDown: (event) => {
+        event.preventDefault();
+      },
+      ArrowUp: (event) => {
+        event.preventDefault();
+      },
+    });
+
+    // TODO: Wire up search
```
Comment (jonator) [inline packages/web/components/drawers/token-select-drawer-limit.tsx:192]: We have a `useFilteredData` hook that can do fuzzy search on client.

File: packages/web/components/input/dynamic-size-input.tsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  value: T;
+  onChange: (newValue: T) => void;
+  className?: string;
+  type: "number" | "string";
+  placeholder?: string;
+  disabled?: boolean;
+  onClick?: () => void;
+}
+
+export const DynamicSizeInput: FC<DynamicSizeInputProps> = ({
```
Comment (jonator) [inline packages/web/components/input/dynamic-size-input.tsx:14]: BTW I think we import a `react-input-autosize` package that may do something like this

File: packages/web/components/place-limit-tool/index.tsx [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    const swapState = usePlaceLimit({
+      osmosisChainId: "localosmosis",
+      poolId,
+      orderDirection,
+      useQueryParams: false,
+      orderbookContractAddress,
+      baseDenom,
+      quoteDenom,
+    });
+    const account = accountStore.getWallet("localosmosis");
```
Comment (jonator) [inline packages/web/components/place-limit-tool/index.tsx:49]: Should we pass the osmosis chain ID? Or is this temporary?

We have an `accountStore.osmosisChainId` member to make this easy

File: packages/web/components/place-limit-tool/index.tsx [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+          </div>
+        </div>
+        <div className="mt-3 flex place-content-between items-center text-body1">
+          <div className="flex w-full flex-col">
+            <div>
+              <span
+                className={classNames(
+                  "w-full bg-transparent text-white-full focus:outline-none"
+                )}
+              >{`When ${swapState.baseDenom} price is at `}</span>
```
Comment (jonator) [inline packages/web/components/place-limit-tool/index.tsx:124]: Let's make sure to not forget localization.
There's some instructions in the web or root README for generating localizations with `yarn machine-translate` in the root of the repo.

File: packages/web/components/place-limit-tool/index.tsx [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+        </div>
+        <Button
+          disabled={swapState.insufficientFunds}
+          isLoading={!swapState.isBalancesFetched}
+          loadingText={"Loading..."}
+          onClick={swapState.placeLimit}
+        >
+          {account?.walletStatus === WalletStatus.Connected ||
+          isSwapToolLoading ? (
+            "Place Order"
```
Comment (jonator) [inline packages/web/components/place-limit-tool/index.tsx:174]: Localization here too

File: packages/web/hooks/limit-orders/use-place-limit.ts [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+    existingAssets: swapAssets.selectableAssets,
+  });
+  console.log("QUOTE", quoteAsset);
+
+  const { asset: baseAsset } = useSwapAsset({
+    minDenomOrSymbol: baseDenom,
+    existingAssets: swapAssets.selectableAssets,
+  });
+
+  const { data: balances, isFetched: isBalancesFetched } = useBalances({
```
Comment (jonator) [inline packages/web/hooks/limit-orders/use-place-limit.ts:111]: Heads up it's not merged yet, but we removed this hook and replaced it with `api.local.balances.getUserBalances` in the `jon/fe-14-lock-tokens` branch (merging soon)

**Review by jonator — COMMENTED** [review]
Initial review

**Comment by jonator** [comment]
Redeployed main build as it seems to have suffered a connection error

File: packages/web/components/place-limit-tool/index.tsx (hunk 14 above)
Comment (crnbarr93) [reply packages/web/components/place-limit-tool/index.tsx:49]: Temporary but will update it to use the `accountStore` reference 👍

File: packages/web/components/control/token-select-limit.tsx [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+                  </span>
+                </div>
+                {orderDirection === OrderDirection.Ask && (
+                  <>
+                    {!isLoadingBasePrice && baseFiatBalance ? (
+                      <div className="flex text-body1 text-osmoverse-300">
+                        {formatPretty(baseFiatBalance)} available
+                      </div>
+                    ) : (
+                      <div />
```
[the remark this answers is not in the document]
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/d647242673e5de8c30ae6700d5d53c65d3dd8ec2

[the remark this answers is not in the document]
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/d647242673e5de8c30ae6700d5d53c65d3dd8ec2

File: packages/web/components/control/token-select-limit.tsx [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+                    {baseAsset.coinDenom}
+                  </span>
+                </div>
+                {orderDirection === OrderDirection.Ask && (
+                  <>
+                    {!isLoadingBasePrice && baseFiatBalance ? (
+                      <div className="flex text-body1 text-osmoverse-300">
+                        {formatPretty(baseFiatBalance)} available
+                      </div>
+                    ) : (
```
[the remark this answers is not in the document]
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/d647242673e5de8c30ae6700d5d53c65d3dd8ec2

File: packages/web/components/control/token-select-limit.tsx (hunk 1 above)
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/ec8948e28a313965a1ccf4618a135b928c1ac119

File: packages/web/components/control/token-select-limit.tsx (hunk 2 above)
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/ec8948e28a313965a1ccf4618a135b928c1ac119

File: packages/web/components/control/token-select-limit.tsx (hunk 3 above)
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/ec8948e28a313965a1ccf4618a135b928c1ac119

File: packages/web/components/control/token-select-limit.tsx (hunk 4 above)
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/ec8948e28a313965a1ccf4618a135b928c1ac119

File: packages/web/components/place-limit-tool/index.tsx (hunk 5 above)
Comment (crnbarr93) [reply packages/web/components/place-limit-tool/index.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/c86e5b34d800f2435163490fac939294643f1333

File: packages/web/components/place-limit-tool/index.tsx (hunk 6 above)
Comment (crnbarr93) [reply packages/web/components/place-limit-tool/index.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/c86e5b34d800f2435163490fac939294643f1333

File: packages/web/components/place-limit-tool/index.tsx (hunk 7 above)
Comment (crnbarr93) [reply packages/web/components/place-limit-tool/index.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/c86e5b34d800f2435163490fac939294643f1333

File: packages/web/components/control/token-select-limit.tsx (hunk 10 above)
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx:165]: This button's been removed in the updated designs, gonna leave it here til we remove it to update for the changes

File: packages/web/components/control/token-select-limit.tsx (hunk 9 above)
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/0c264e9f2c02e4d95002cbd82877722aca38a223

File: packages/web/components/control/token-select-limit.tsx (hunk 8 above)
Comment (crnbarr93) [reply packages/web/components/control/token-select-limit.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/9d8f4395e62ff823a46bdc59fb3838b80b4cdb9c

File: packages/web/components/drawers/token-select-drawer-limit.tsx (hunk 12 above)
Comment (crnbarr93) [reply packages/web/components/drawers/token-select-drawer-limit.tsx:192]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/1a3e662ad0263660bb65f96c479f75a1c6a1df35

File: packages/web/components/place-limit-tool/index.tsx (hunk 15 above)
Comment (crnbarr93) [reply packages/web/components/place-limit-tool/index.tsx:124]: This looks like it's been changed in design so will hold off on this one, will update others

File: packages/web/components/place-limit-tool/index.tsx (hunk 14 above)
Comment (crnbarr93) [reply packages/web/components/place-limit-tool/index.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/c8466446eaec7a82b02e6f4bbc2c2b614c1901b8

File: packages/web/components/place-limit-tool/index.tsx (hunk 16 above)
Comment (crnbarr93) [reply packages/web/components/place-limit-tool/index.tsx]: https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/c4a880322b2c3472d38898483f6a8c304cc7fa67

File: packages/web/hooks/limit-orders/use-place-limit.ts (hunk 17 above)
Comment (crnbarr93) [reply packages/web/hooks/limit-orders/use-place-limit.ts:111]: Mind shooting me a message when this is merged?

File: packages/web/components/drawers/token-select-drawer-limit.tsx (hunk 11 above)
Comment (crnbarr93) [reply packages/web/components/drawers/token-select-drawer-limit.tsx:82]: Nothing stands out immediately but maybe a way to repurpose one of the other hooks?

This is the same as `token-select-drawer`,  Ieft this part unedited.

**Pull request merged**